sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -681,7 +681,9 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
           if supportedUnaryExpression(u) && atMostOneUnfoldable(branches.map(_._2) ++ elseValue) =>
         c.copy(
           branches.map(e => e.copy(_2 = u.withNewChildren(Array(e._2)))),
-          elseValue.map(e => u.withNewChildren(Array(e))))
+          elseValue
+            .orElse { if (u.isInstanceOf[IsNull]) Some(Literal(null, c.dataType)) else None }
```

Comment:
Wouldn't it be safer to allways return `Some(Literal(null, c.dataType))` here?

Reply:
Yes. It should be safer, but it will add a useless `Literal(null, c.dataType)` in most cases. Another way is:
```
elseValue.orElse(if (u.isInstanceOf[Predicate]) Some(Literal(null, c.dataType)) else None)
```

Reply:
Can we optimize away the useless else branch? I think null propagation + removing null else branch should do the work?